modApiServer/src/org/aion/api/server/http/ApiWeb3Aion.java

```diff
@@ -154,293 +159,724 @@ AionBlock getBlockRaw(int bn) {
+
+        String bnOrId = "latest";
+        if (_bnOrId != null && !_bnOrId.equals(null))
+            bnOrId = _bnOrId + "";
+
+        BigInteger balance = getRepoByJsonBlockId(bnOrId).getBalance(address);
+        return TypeConverter.toJsonHex(balance);
+    }
+
+    public Object eth_getStorageAt(String _address, String _storageIndex, Object _bnOrId) {
+        Address address = new Address(_address);
+
+        String bnOrId = "latest";
+        if (_bnOrId != null && !_bnOrId.equals(null))
+            bnOrId = _bnOrId + "";
+
+        DataWord key = DataWord.ZERO;
 
+        try {
+            key = new DataWord(ByteUtil.hexStringToBytes(_storageIndex));
+        } catch (Exception e) {
+            // invalid key
+            LOG.debug("eth_getStorageAt: invalid storageIndex. Must be <= 16 bytes.");
+            return null;
+        }
+
+        DataWord storageValue = (DataWord) getRepoByJsonBlockId(bnOrId).getStorageValue(address, key);
+        return storageValue != null ? TypeConverter.toJsonHex(storageValue.getData()) : null;
+    }
+
+    public Object eth_getTransactionCount(String _address, Object _bnOrId) {
+        Address address = new Address(_address);
+
+        String bnOrId = "latest";
+        if (_bnOrId != null && !_bnOrId.equals(null))
+            bnOrId = _bnOrId + "";
+
+        return TypeConverter.toJsonHex(getRepoByJsonBlockId(bnOrId).getNonce(address));
+    }
+
+    public Object eth_getBlockTransactionCountByHash(String hashString) {
         byte[] hash = ByteUtil.hexStringToBytes(hashString);
+        AionBlock b = this.ac.getBlockchain().getBlockByHash(hash);
+        if (b == null) return null;
+        long n = b.getTransactionsList().size();
+        return TypeConverter.toJsonHex(n);
+    }
+
+    public Object eth_getBlockTransactionCountByNumber(String _bnOrId) {
+        List<AionTransaction> list = getTransactionsByBlockId(_bnOrId);
+        if (list == null) return null;
+        long n = list.size();
+        return TypeConverter.toJsonHex(n);
+    }
+
+    public Object eth_getCode(String _address, Object _bnOrId) {
+        Address address = new Address(_address);
+
+        String bnOrId = "latest";
+        if (_bnOrId != null && !_bnOrId.equals(null))
+            bnOrId = _bnOrId + "";
+
+        IRepository repo = getRepoByJsonBlockId(bnOrId);
+        if (repo == null) return null; // invalid bnOrId
+
+        byte[] code = repo.getCode(address);
+        return TypeConverter.toJsonHex(code);
+    }
+
+    public Object eth_sign(String _address, String _message) {
+        Address address = Address.wrap(_address);
+        ECKey key = getAccountKey(address.toString());
+        if (key == null)
+            return null;
+
+        // Message starts with Unicode Character 'END OF MEDIUM' (U+0019)
+        String message = "\u0019Aion Signed Message:\n" + _message.length() + _message;
+        byte[] messageHash = HashUtil.keccak256(message.getBytes());
```

Comment:
Any reason for using keccak256?

Reply:
Ethereum compatibility. We can change this in the future once we decide on a message signature format. Note: The Ethereum message signature has several incompatible implementations due to poor initial spec.